File: Validation/vali.php
<?php

declare(strict_types=1);

final class Validator
{
    // 正規表現（日本向け実用パターン）
    private const PHONE_REGEX       = '/^0\d{1,4}[-\s]?\d{1,4}[-\s]?\d{4}$|^0\d{9,10}$/';
    private const POSTAL_CODE_REGEX = '/^\d{3}-?\d{4}$/';
    private const PASSWORD_REGEX    = '/^(?=.*[a-z])(?=.*[A-Z])(?=.*\d).{8,}$/';

    /**
     * メールアドレス検証（filter_var で十分安全）
     */
    public static function email(string $email): bool
    {
        return filter_var($email, FILTER_VALIDATE_EMAIL) !== false;
    }

    /**
     * 日本国内電話番号（ハイフン・スペースあり/なし、携帯もOK）
     */
    public static function phone(string $phone): bool
    {
        $cleaned = preg_replace('/[^\d]/', '', $phone);
        return preg_match(self::PHONE_REGEX, $phone) === 1
            || preg_match('/^0\d{9,10}$/', $cleaned) === 1;
    }

    /**
     * パスワード強度（8文字以上 + 大文字・小文字・数字）
     */
    public static function password(string $password): bool
    {
        return preg_match(self::PASSWORD_REGEX, $password) === 1;
    }

    /**
     * 日付検証（デフォルト Y-m-d）
     */
    public static function date(string $date, string $format = 'Y-m-d'): bool
    {
        $d = DateTime::createFromFormat('!' . $format, $date);
        return $d !== false && $d->format($format) === $date;
    }

    /**
     * 日本の郵便番号（123-4567 または 1234567）
     */
    public static function postalCode(string $postalCode): bool
    {
        return preg_match(self::POSTAL_CODE_REGEX, trim($postalCode)) === 1;
    }

    /**
     * クレジットカード番号（Luhnアルゴリズム）
     */
    public static function creditCard(string $cardNumber): bool
    {
        $digits = preg_replace('/\D/', '', $cardNumber);

        if (strlen($digits) < 13 || strlen($digits) > 19) {
            return false;
        }

        $sum = 0;
        $reverse = array_reverse(str_split($digits));

        foreach ($reverse as $i => $digit) {
            $value = (int)$digit;
            if ($i % 2 === 1) {
                $value *= 2;
                if ($value > 9) {
                    $value -= 9;
                }
            }
            $sum += $value;
        }

        return $sum % 10 === 0;
    }

    /**
     * ユーザー登録フォームなど一括バリデーション
     *
     * @param array $data
     * @param array $rules カスタムルール指定可能（デフォルトは一般的な登録フォーム用）
     * @return array{is_valid: bool, errors: array<string>}
     */
    public static function form(array $data, array $rules = []): array
    {
        // デフォルトルール（必要に応じて上書き可能）
        $defaultRules = [
            'email'       => ['required' => true,  'rule' => 'email'],
            'phone'       => ['required' => false, 'rule' => 'phone'],
            'password'    => ['required' => true,  'rule' => 'password'],
            'birth_date'  => ['required' => false, 'rule' => 'date'],
            'postal_code' => ['required' => false, 'rule' => 'postalCode'],
        ];

        $rules = array_merge($defaultRules, $rules);
        $errors = [];

        foreach ($rules as $field => $config) {
            $value = $data[$field] ?? null;
            $label = $config['label'] ?? ucwords(str_replace('_', ' ', $field));

            // 必須チェック
            if ($config['required'] && (empty($value) && $value !== '0')) {
                $errors[] = "{$label}は必須です。";
                continue;
            }

            // 値が空なら以降の検証はスキップ
            if (empty($value) && $value !== '0') {
                continue;
            }

            // 個別ルール検証
            $valid = match ($config['rule']) {
                'email'       => self::email($value),
                'phone'       => self::phone($value),
                'password'    => self::password($value),
                'date'        => self::date($value),
                'postalCode'  => self::postalCode($value),
                'creditCard'  => self::creditCard($value),
                default       => true,
            };

            if (!$valid) {
                $errors[] = $config['message'] ?? self::defaultErrorMessage($field, $config['rule']);
            }
        }

        return [
            'is_valid' => empty($errors),
            'errors' => $errors
        ];
    }

    /**
     * デフォルトエラーメッセージ
     */
    private static function defaultErrorMessage(string $field, string $rule): string
    {
        return match ($rule) {
            'email'       => '正しいメールアドレスを入力してください。',
            'phone'       => '電話番号の形式が正しくありません（例: 090-1234-5678 または 09012345678）',
            'password'    => 'パスワードは8文字以上で、大文字・小文字・数字をそれぞれ1文字以上含めてください。',
            'date'        => '日付の形式が正しくありません（例: 2000-01-31）',
            'postalCode'  => '郵便番号は 123-4567 または 1234567 の形式で入力してください。',
            'creditCard'  => 'クレジットカード番号が正しくありません。',
            default       => '入力内容に誤りがあります。',
        };
    }
}

// ──────────────────────────────────────────────────
// 使用例
// ──────────────────────────────────────────────────
if (php_sapi_name() === 'cli' || basename($_SERVER['SCRIPT_FILENAME']) === basename(__FILE__)) {

    $userData = [
        'email'       => 'user@example.co.jp',
        'phone'       => '090-1234-5678',        // OK
        // 'phone'    => '09012345678',          // これもOK
        'password'    => 'MyStrongPass123',
        'birth_date'  => '1995-05-20',
        'postal_code' => '1000001',              // ハイフンなしもOK
    ];

    $result = Validator::form($userData);

    if (!$result['is_valid']) {
        echo "検証エラー:\n";
        foreach ($result['errors'] as $error) {
            echo "・ {$error}\n";
        }
    } else {
        echo "すべての入力が正しいです！\n";
    }

    // クレジットカード単体テスト例
    var_dump(Validator::creditCard('4111-1111-1111-1111')); // bool(true)
}

File: Validation/vali.java
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.*;
import java.util.regex.Pattern;

public final class DataValidator {

    // 正規表現（コンパイル済みで高速）
    private static final Pattern EMAIL_PATTERN = Pattern.compile(
        "^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,]$"
    );

    // 日本電話番号：ハイフン・スペースあり/なし、携帯・固定・IP電話対応
    private static final Pattern PHONE_PATTERN = Pattern.compile(
        "^0\\d{1,4}[-\\s]?\\d{1,4}[-\\s]?\\d{4}$|^0\\d{9,10}$"
    );

    // 郵便番号：123-4567 または 1234567
    private static final Pattern POSTAL_CODE_PATTERN = Pattern.compile(
        "^\\d{3}-?\\d{4}$"
    );

    // パスワード：8文字以上 + 大文字・小文字・数字をそれぞれ1文字以上
    private static final Pattern PASSWORD_PATTERN = Pattern.compile(
        "^(?=.*[a-z])(?=.*[A-Z])(?=.*\\d).{8,}$"
    );

    private DataValidator() {} // インスタンス化防止

    /** メールアドレス検証 */
    public static boolean isValidEmail(String email) {
        return email != null && EMAIL_PATTERN.matcher(email).matches();
    }

    /** 電話番号検証（日本向け） */
    public static boolean isValidPhone(String phone) {
        if (phone == null) return false;
        String cleaned = phone.replaceAll("[\\s-]", "");
        return PHONE_PATTERN.matcher(phone).matches() || 
               PHONE_PATTERN.matcher(cleaned).matches();
    }

    /** パスワード強度検証 */
    public static boolean isValidPassword(String password) {
        return password != null && PASSWORD_PATTERN.matcher(password).matches();
    }

    /** 日付検証（デフォルト yyyy-MM-dd） */
    public static boolean isValidDate(String dateStr) {
        return isValidDate(dateStr, "yyyy-MM-dd");
    }

    public static boolean isValidDate(String dateStr, String pattern) {
        if (dateStr == null) return false;
        try {
            DateTimeFormatter formatter = DateTimeFormatter.ofPattern(pattern);
            LocalDate.parse(dateStr, formatter);
            return true;
        } catch (DateTimeParseException e) {
            return false;
        }
    }

    /** 郵便番号検証（日本） */
    public static boolean isValidPostalCode(String postalCode) {
        return postalCode != null && POSTAL_CODE_PATTERN.matcher(postalCode.trim()).matches();
    }

    /** クレジットカード番号検証（Luhnアルゴリズム） */
    public static boolean isValidCreditCard(String cardNumber) {
        if (cardNumber == null) return false;

        String digits = cardNumber.replaceAll("\\D", "");
        if (digits.length() < 13 || digits.length() > 19) return false;

        int sum = 0;
        boolean shouldDouble = true; // 右から2番目から2倍

        for (int i = digits.length() - 1; i >= 0; i--) {
            int digit = digits.charAt(i) - '0';

            if (shouldDouble) {
                digit *= 2;
                if (digit > 9) digit -= 9;
            }

            sum += digit;
            shouldDouble = !shouldDouble;
        }

        return sum % 10 == 0;
    }

    // ===================================================================
    // 柔軟なフォームバリデーション（ルールベース）
    // ===================================================================

    public record ValidationError(String field, String message) {}

    public record ValidationResult(boolean isValid, List<ValidationError> errors) {
        public static final ValidationResult SUCCESS = new ValidationResult(true, List.of());

        public ValidationResult {
            errors = List.copyOf(errors); // 不変リストに
        }

        public void throwIfInvalid() {
            if (!isValid) {
                throw new IllegalArgumentException("Validation failed: " + errors);
            }
        }
    }

    @FunctionalInterface
    public interface ValidatorRule {
        Optional<String> validate(String value);
    }

    public static ValidationResult validate(Map<String, String> data, Map<String, ValidatorRule> rules) {
        List<ValidationError> errors = new ArrayList<>();

        for (var entry : rules.entrySet()) {
            String field = entry.getKey();
            String value = data.getOrDefault(field, "").trim();
            ValidatorRule rule = entry.getValue();

            if (value.isEmpty()) {
                errors.add(new ValidationError(field, fieldLabel(field) + "は必須です。"));
                continue;
            }

            rule.validate(value).ifPresent(msg ->
                errors.add(new ValidationError(field, msg))
            );
        }

        return errors.isEmpty()
            ? ValidationResult.SUCCESS
            : new ValidationResult(false, errors);
    }

    // 便利なプリセットルール
    public static final Map<String, ValidatorRule> USER_FORM_RULES = Map.of(
        "email",       v -> isValidEmail(v)       ? Optional.empty() : Optional.of("正しいメールアドレスを入力してください。"),
        "phone",       v -> isValidPhone(v)       ? Optional.empty() : Optional.of("電話番号の形式が正しくありません（例: 090-1234-5678 または 09012345678）"),
        "password",    v -> isValidPassword(v)    ? Optional.empty() : Optional.of("パスワードは8文字以上で、大文字・小文字・数字をそれぞれ1文字以上含めてください。"),
        "birth_date",  v -> isValidDate(v)        ? Optional.empty() : Optional.of("生年月日は yyyy-MM-dd 形式で入力してください（例: 2000-01-31）"),
        "postal_code", v -> isValidPostalCode(v)  ? Optional.empty() : Optional.of("郵便番号は 123-4567 または 1234567 の形式で入力してください。")
    );

    private static String fieldLabel(String field) {
        return switch (field) {
            case "email"       -> "メールアドレス";
            case "phone"       -> "電話番号";
            case "password"    -> "パスワード";
            case "birth_date"  -> "生年月日";
            case "postal_code" -> "郵便番号";
            default            -> field;
        };
    }

    // ===================================================================
    // 使用例
    // ===================================================================
    public static void main(String[] args) {
        var validator = new DataValidator();

        Map<String, String> userData = Map.of(
            "email",       "taro@example.co.jp",
            "phone",       "090-1234-5678",      // OK
            // "phone",    "09012345678",        // これもOK
            "password",    "SecurePass123",
            "birth_date",  "1990-05-20",
            "postal_code", "1000001"             // ハイフンなしOK
        );

        ValidationResult result = validator.validate(userData, USER_FORM_RULES);

        if (result.isValid()) {
            System.out.println("すべての入力が正しいです！");
        } else {
            System.out.println("入力エラー:");
            result.errors().forEach(e ->
                System.out.println("・ " + e.field() + ": " + e.message())
            );
        }

        // 単体テスト例
        System.out.println("\nクレジットカードテスト:");
        System.out.println("4111111111111111 → " + isValidCreditCard("4111-1111-1111-1111")); // true
        System.out.println("1234567890123456 → " + isValidCreditCard("1234567890123456")); // false
    }
}